baseapp/grpcrouter.go

```diff
@@ -0,0 +1,112 @@
+package baseapp
+
+import (
+	gocontext "context"
+	"fmt"
+
+	gogogrpc "github.com/gogo/protobuf/grpc"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/encoding/proto"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var protoCodec = encoding.GetCodec(proto.Name)
+
+// GRPCRouter routes ABCI Query requests to GRPC handlers
+type GRPCRouter struct {
+	routes map[string]GRPCQueryHandler
+}
+
+var _ gogogrpc.Server
+
+// NewGRPCRouter creates a new GRPCRouter
+func NewGRPCRouter() *GRPCRouter {
+	return &GRPCRouter{
+		routes: map[string]GRPCQueryHandler{},
+	}
+}
+
+// GRPCQueryHandler defines a function type which handles ABCI Query requests
+// using gRPC
+type GRPCQueryHandler = func(ctx sdk.Context, req abci.RequestQuery) (abci.ResponseQuery, error)
+
+// Route returns the Querier for a given query route path.
+func (qrt *GRPCRouter) Route(path string) GRPCQueryHandler {
+	handler, found := qrt.routes[path]
+	if !found {
+		return nil
+	}
+	return handler
+}
+
+// RegisterService implements the grpc Server.RegisterService method
```

Comment:
The godoc should describe what the arguments are and how they're used. In addition, it should detail what is returned. Finally, also touch up on methodHandler and what `res` is returned on that.